Port Compression Streams to Apple Compression Framework

stwrt opened the pull request:
#### 92a21241fd9b82ff695f130feb2b77326556fe3c
<pre>
Port Compression Streams to Apple Compression Framework
<a href="https://bugs.webkit.org/show_bug.cgi?id=244009">https://bugs.webkit.org/show_bug.cgi?id=244009</a>

Reviewed by Youenn Fablet.

This change ports the decompression deflate-raw to use the Apple Compression Framework.
All Compression and Decompression using Deflate or Gzip will still rely on Zlib, and all other
platforms will continue using Zlib.

Syncing WPT test changes from <a href="https://github.com/web-platform-tests/wpt/pull/35560">https://github.com/web-platform-tests/wpt/pull/35560</a>

* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.js:
(const.chunk.of.badChunks.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.serviceworker-expected.txt:
* Source/WebCore/Modules/compression/DecompressionStream.js:
(initializeDecompressionStream):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
(WebCore::DecompressionStreamDecoder::decode):
(WebCore::DecompressionStreamDecoder::flush):
(WebCore::DecompressionStreamDecoder::initialize):
(WebCore::DecompressionStreamDecoder::decompressZlib):
(WebCore::DecompressionStreamDecoder::initializeAppleCompressionFramework):
(WebCore::DecompressionStreamDecoder::decompressAppleCompressionFramework):
(WebCore::DecompressionStreamDecoder::decompress): Deleted.
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
(WebCore::DecompressionStreamDecoder::~DecompressionStreamDecoder):
(WebCore::DecompressionStreamDecoder::DecompressionStreamDecoder):

Canonical link: <a href="https://commits.webkit.org/253697@main">https://commits.webkit.org/253697@main</a>
</pre>







<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/baeedf1bf002c399b63f747ca67fa3bbd4f2d566

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86825 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30898 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17714 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95657 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149409 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/90805 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29268 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25634 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78986 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90892 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92441 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23638 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/73707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23676 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78634 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/78938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66681 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27029 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/12791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26951 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/13805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2622 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28635 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/73707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28579 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/33084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
```diff
@@ -64,14 +74,35 @@ class DecompressionStreamDecoder : public RefCounted<DecompressionStreamDecoder>
     bool finish { false };
     z_stream zstream;
 
+    bool usingAppleCompressionFramework { false };
```
All member variables should start with m_, including stream, finish, stream...
Can you update DecompressionStreamDecoder (and Encoder if needed), in a  separate PR probably.

stwrt replied:
Thanks will fix

stwrt replied:
Should function pointers also have `m_` in front?

`ExceptionOr<Vector<uint8_t>> (DecompressionStreamDecoder::*decompress)(const uint8_t*, const size_t);`

youennf replied:
I would guess so

youennf commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
```diff
@@ -25,6 +25,10 @@
 #pragma once
 
 #include "BufferSource.h"
+#ifdef __APPLE__
```
We do not use __APPLE__ a lot. Can we use #if PLATFORM(COCOA) instead?

stwrt replied:
Ah, was not aware of that. Going to update this.

youennf replied:
There are still some __APPLE__ in the current PR

youennf commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -147,4 +143,72 @@ ExceptionOr<Vector<uint8_t>> DecompressionStreamDecoder::decompress(const uint8_
+{
+    size_t index = 0, allocateSize = startingAllocationSize, totalSize = 0;
+    Vector<Vector<uint8_t>>storage;
+    storage.reserveCapacity(128);
```
I am not sure reserveCapacity is needed here.

stwrt replied:
I was trying to reserve some space up front, so we don't need to keep reallocating the Vector of pointers.

I was using the reserveCapacity, so it didn't modify the size of vector just the capacity.

stwrt replied:
Rewrote logic. Above comment does not apply anymore 😀

youennf commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -147,4 +143,72 @@ ExceptionOr<Vector<uint8_t>> DecompressionStreamDecoder::decompress(const uint8_
+    Vector<uint8_t> output;
+    output.reserveCapacity(totalSize);
+    for (auto& storageElement : storage)
+        output.append(storageElement.data(), storageElement.size());
```
You might want to consider using SharedBufferBuilder instead of having a Vector of Vector<uint8_t>, then call extractData().

stwrt replied:
Will investigate this today

stwrt replied:
Thanks for the recommendation for `SharedBufferBuilder`. It simplified quite a bit of the logic here.

I should port it over to CompressionStreamEncoder in another patch.

youennf commented on LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.js:
```diff
@@ -56,8 +56,8 @@ for (const chunk of badChunks) {
-    await promise_rejects_js(t, TypeError, writePromise, 'write should reject');
-    await promise_rejects_js(t, TypeError, readPromise, 'read should reject');
+    await promise_rejects_js(t, TypeError, writer.close(), 'write should reject');
+    await promise_rejects_js(t, TypeError, reader.closed, 'read should reject');
```
It is suspicious that we need to change the test while we are only switching from zlib to Apple Compression framework.
It might be due to lib eagerly doing some work and failing early while Apple Compression format is lazily doing things.
Can we understand what is going on here? Is there a way to use Apple Compression framework while keeping the exact zlib behavior with regards to error reporting?

stwrt replied:
The "invalid gzip bytes" test is throwing an error early in Zlib during the decompress stage. The compression framework on the other hand waits to error until we do the flush. Based on internal discussions with the Compression Framework team they did not indicate this was an issue.

We should not be tying this api exclusively to the Zlib behavior though. If compression framework decides to throw the error later we should support it.

youennf replied:
This seems ok, we can always revisit if there are too many behavioural differences.
Please add some comments explaining why you are changing the test though in the commit message.
Please also do a PR on the WPT repository with those changes as otherwise, these changes would get lost next time we resync WPT tests.

geoffreygaren replied:
@stwrt This change is probably fine, if you follow @youennf's guidance to update WPT. But I want to comment about how we frame a discussion like this.

Talking to the Compression Framework team is of course a good practice. But when we make decisions about the specifics of WebKit's behavior, our rationale should be based on what the web standard says, what other browsers do, and what web developers might reasonably expect, rather than what happens by default in our underlying frameworks. If an underlying framework does not match standards and/or other browsers and/or web developer expectations by default, it's our job to write the glue code that bridges that gap. If an underlying framework does not match standards and/or other browsers and/or web developer expectations for a Very Good Reason (TM), it's our job to influence change in standards and other browsers to conform to that Very Good Reason (TM).

stwrt replied:
I'll make sure to add a comment and post the change to WPT.

youennf commented on LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.serviceworker-expected.txt:
```diff
@@ -1,4 +1,6 @@
 
+Harness Error (FAIL), message = Unhandled rejection: Failed to Decode Data.
```
Which promise is actually not handled? readPromise maybe?
Can we change the test to not have this message?

stwrt replied:
Going to take a look into that

stwrt replied:
The most interesting thing I noticed is that this does not trigger in the other mutations (non service worker, shared worker, or worker). I wonder what makes service worker special.

stwrt replied:
Still need to fix

youennf replied:
I think we should write the tests in this way:
1. writer.close() should be called and should reject so it is good to add the promise_rejects_js. Ditto for reader.closed.
2. writePromise may not always reject as we have seen, but we need to handle the promise in case it is rejecting. something like writePromise.then(() => { }, () => { });
3. reader.read() should probably reject. If it does not reject, it will mean we will have a bigger behavioural difference since some potentially bad data will be given to the application. If it always reject, we should add a promise_rejects_js assertion. Otherwise do then(() => {}, () => {}).

stwrt replied:
I rewrote the tests to follow this pattern.

It would be worth having a discussion around the differences of when the error is thrown between the two frameworks, and what a developer's expectation would be in this case.

youennf commented on Source/WebCore/Modules/compression/DecompressionStream.js:
```diff
@@ -68,20 +68,19 @@ function initializeDecompressionStream(format)
         try {
-        buffer = decoder.@flush();
+            const decoder = @getByIdDirectPrivate(this, "DecompressionStreamDecoder");
+            let buffer = decoder.@flush();
```
could be `const buffer`

youennf commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -38,50 +38,46 @@ ExceptionOr<RefPtr<Uint8Array>> DecompressionStreamDecoder::decode(const BufferS
-    zstream.opaque = 0;
-    zstream.zalloc = 0;
-    zstream.zfree = 0;
+    m_initailized = true;
```
should be m_initialized.

youennf commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
```diff
@@ -60,18 +72,39 @@ class DecompressionStreamDecoder : public RefCounted<DecompressionStreamDecoder>
     explicit DecompressionStreamDecoder(unsigned char format) 
         : m_format(static_cast<Formats::CompressionFormat>(format))
     {
+        decompress = &DecompressionStreamDecoder::decompressZlib;
```
I am not sure it is worth having a function pointer here.
This is totally unneeded for non COCOA ports.
And for Cocoa ports, we could simply introduce a decompress method that would check m_formatto either call decompressAppleCompressionFramework or decompressZlib.
Code would be easier to read and I do not think this will change much in efficiency.

stwrt replied:
I was debating whether to use function pointers or not.

The code in `decode()` and `flush()` got a lot uglier to read without it.

You would ultimately need to do something like below...

```CPP
RefPtr<JSC::ArrayBuffer> compressedDataCheck;

#if PLATFORM(COCOA)
if(m_usingAppleCompressionFramework)
   compressedDataCheck = decompressAppleCompressionFramework(data, input.length());
else
   compressedDataCheck = decompressZlib(data, input.length());
#else
   compressedDataCheck = decompressZlib(data, input.length());
#endif
```

I also did try a few other methods of abstraction, but function pointers came out as the least ugly to read at the end of the day.

Function pointers allow us to hide all the complex logic in the initializer, so `decode()` and `flush()` read cleanly.

I agree with the point that function pointers serve little purpose on non Apple platforms.

stwrt replied:
Got rid of the function ptrs in a clean way.

youennf commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
```diff
@@ -60,18 +72,36 @@ class DecompressionStreamDecoder : public RefCounted<DecompressionStreamDecoder>
     explicit DecompressionStreamDecoder(unsigned char format) 
         : m_format(static_cast<Formats::CompressionFormat>(format))
     {
+        std::memset(&m_zstream, 0, sizeof(m_zstream));
```
Probably worth moving to cpp file

stwrt replied:
Thanks. Moved both `memset`s to the initialize functions.

youennf commented on LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.js:
```diff
@@ -66,8 +68,10 @@ for (const chunk of badChunks) {
+    await promise_rejects_js(t, TypeError, writer.close(), 'write should reject');
+    await promise_rejects_js(t, TypeError, reader.closed, 'read should reject');
+    writePromise.then(() => {}, () => {});
+    readPromise.then(() => {}, () => {});
```
I would just do a one liner instead (maybe adding a comment about error handling):
writer.write(chunk.value).then(() => {}, () => {});
reader.read().then(() => {}, () => {});

Once we have this reader.read(), maybe we can add a reader.read() rejection, called after writer.close().
Something like `await promise_rejects_js(t, TypeError, reader.read(), 'write should reject');`

stwrt replied:
That sounds good.

youennf reviewed (approved):
Looks good. Let's merge it once we have the WPT changes landed in WPT repo.

stwrt committed:
Port Compression Streams to Apple Compression Framework
https://bugs.webkit.org/show_bug.cgi?id=244009

Reviewed by Youenn Fablet.

This change ports the decompression deflate-raw to use the Apple Compression Framework.
All Compression and Decompression using Deflate or Gzip will still rely on Zlib, and all other
platforms will continue using Zlib.

Syncing WPT test changes from https://github.com/web-platform-tests/wpt/pull/35560

* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.js:
(const.chunk.of.badChunks.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.serviceworker-expected.txt:
* Source/WebCore/Modules/compression/DecompressionStream.js:
(initializeDecompressionStream):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
(WebCore::DecompressionStreamDecoder::decode):
(WebCore::DecompressionStreamDecoder::flush):
(WebCore::DecompressionStreamDecoder::initialize):
(WebCore::DecompressionStreamDecoder::decompressZlib):
(WebCore::DecompressionStreamDecoder::initializeAppleCompressionFramework):
(WebCore::DecompressionStreamDecoder::decompressAppleCompressionFramework):
(WebCore::DecompressionStreamDecoder::decompress): Deleted.
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
(WebCore::DecompressionStreamDecoder::~DecompressionStreamDecoder):
(WebCore::DecompressionStreamDecoder::DecompressionStreamDecoder):

Canonical link: https://commits.webkit.org/253697@main
